common/parallel/FileSystemCrawler.go

```diff
@@ -132,18 +145,23 @@ func enumerateOneFileSystemDirectory(dir Directory, enqueueDir func(Directory),
 			}
 			return nil
 		} else if err != nil {
-			return err
+			// FileInfo value being nil should mean that the FileSystemEntry refers to a directory.
+			enqueueOutput(FileSystemEntry{dirString, nil}, err)
+
+			// Since we have already enqueued the failed enumeration entry, return nil error to avoid duplicate queueing by workerLoop().
+			return nil
 		}
 		for _, childInfo := range list {
-			if failable, ok := childInfo.(failableFileInfo); ok && failable.Error() != nil {
-				// while Readdir as a whole did not fail, this particular file info did
-				enqueueOutput(FileSystemEntry{}, failable.Error())
-				continue
-			}
 			childEntry := FileSystemEntry{
 				fullPath: filepath.Join(dirString, childInfo.Name()),
 				info:     childInfo,
 			}
+
+			if failable, ok := childInfo.(failableFileInfo); ok && failable.Error() != nil {
+				// while Readdir as a whole did not fail, this particular file info did
+				enqueueOutput(childEntry, failable.Error())
+				continue
+			}
```

Comment:
QQ: Why did we shift this segment from above?

Reply:
We need the childEntry as an argument to the enqueueOutput() method. Hence, the segment has been shifted.